### ISSUE TITLE
`volta install corepack` doesn't provide corepack's `yarn`

### ISSUE THREAD

**Issue by gamoutatsumi** [volta-cli/volta#987]
I installed https://github.com/nodejs/corepack.
Normally corepack provide yarn binary, but using volta, yarn is linked to `volta-shim`.
Volta has not compatibility for corepack?

**Comment by charlespierce** [comment]
Hi @gamoutatsumi, unfortunately, I think it's true that Volta isn't compatible with globally installing corepack. Ultimately, both Volta and corepack are looking to solve the same problem: Making a package manager available to you without relying on a specific version being installed.

Volta's has special-case handling for the package managers, so when you run `yarn` it uses Volta's logic for detecting the package manager, as opposed to delegating to Corepack.

**Comment by ljharb** [comment]
Given that node is planning to integrate corepack by default, you may want to consider supporting it.

**Comment by charlespierce** [comment]
Yeah, I think we'll need to evaluate how to do a partial passthrough for when Node starts shipping Corepack. We need to do it in a way that won't break down on older Node versions that _don't_ include Corepack. I don't know that we can reasonably support the 3rd-party package version without a significant refactor, however, and once it's included by default the value-add is quite a bit lower.

**Comment by eugene1g** [comment]
`corepack` is now bundled with Node as of `16.9.0` (https://github.com/nodejs/node/blob/master/doc/changelogs/CHANGELOG_V16.md#corepack) but not yet enabled by default. 

Today we use Volta to control versions for both Node and Yarn, but we now plan to use Volta for Node (in dev / prod), and `corepack` for Yarn (in dev). Does Volta plan to support this dual proxying, or is that getting too complex?

**Comment by cspotcode** [comment]
I was able to make this work as follows:

```
npm install -g corepack

# Specifying an explicit install-directory makes corepack overwrite volta's yarn shims, which is what we want
corepack enable --install-directory ~/.volta/bin

# Now yarn should be controlled by corepack
yarn set version berry
```

**Comment by charlespierce** [comment]
Looking at how we handle the passthrough behavior, I think that we can make Volta interop with Corepack nicely with a minimal change. It should work as it currently does for Node without Corepack (showing a nice error about missing Yarn), while with Corepack it will delegate to that behavior if you run `yarn`.

**Comment by sirreal** [comment]
Corepack is bundled with recent node versions. The [documentation](https://nodejs.org/api/corepack.html) suggests it's available as an executable just like `node`:

> … run `corepack enable`
> … run `corepack prepare`

It's experimental at the moment, but given Corepack is expected to be core functionality, when installing a compatible version of node should volta shim `corepack` to the bundled version as well?

**Comment by charlespierce** [comment]
@sirreal Yeah, once we have support for calling the Corepack version of `yarn` (Started in #1024, got blocked on showing a proper error message so will need some design work), I think we'll also need to add a shim for `corepack` itself. That should be relatively straightforward, and use the same logic as the `npx` executable, showing a helpful error message if you're on an older version of Node that doesn't include `corepack`.

**Comment by brandonchinn178** [comment]
IIUC, the only benefit of using corepack's yarn is because corepack looks at the `packageManager` key in `package.json` and automatically uses the right version of `yarn` (instead of checking in the yarn binary). In lieu of getting some sort of compatibility between volta and corepack, wouldn't it be sufficient to update `volta` to support looking at `packageManager` and using the same logic?

**Comment by charlespierce** [comment]
@brandonchinn178 Yeah, that definitely would be an intermediate approach to help improve compatibility. There's a bit of design that would need to happen there (we'd definitely love a small RFC!) around how we handle `volta pin` when the version is pulled from `packageManager` instead of `volta`.

Somewhat related to #282, which tracks the potential for pulling version information from different sources.

**Comment by devinrhode2** [comment]
Deprecate `volta` inside package.json and use the new standard `packageManger` or whatever

**Comment by devinrhode2** [comment]
If we can avoid adding "branded" metadata into package.json, more people can have more freedom wrt their choice of node manager

**Comment by devinrhode2** [comment]
Some teams don't like adding metadata for individual developers unique tooling choices :)

**Comment by charlespierce** [comment]
Hi @devinrhode2, that's definitely an option we could investigate. The issue with that is it doesn't support choosing the Node version, which is necessary for Volta. We're certainly sympathetic to teams / projects that don't want to accommodate a myriad of ways to support the different developer tooling. We have another issue #282 focused specifically around supporting other mechanisms for defining the tool versions. The main blocker is there are a _lot_ of edge cases and nuances that would need to be worked out before we can move forward.

**Comment by 0xdevalias** [comment]
Was just wondering if any more progress had been made on potentially landing this PR?:

- https://github.com/volta-cli/volta/issues/282

---

> Deprecate `volta` inside `package.json` and use the new standard `packageManger`

This would be my suggestion as well, at least for the aspects that `corepack` and that `packageManager` field cover. 

Probably done as an opt-in release of `volta` at first, that lets users elect to read from the `packageManager` versions. If that is opted into, then it should through at least a warning (maybe an error) if versions are set in the old `volta` keys, and should tell the user to set/update them in the `packageManager` keys.

Then after some time as deemed appropriate, a new 'breaking change' release of `volta` could be made that switches this to the default mode.

> The issue with that is it doesn't support choosing the Node version, which is necessary for Volta. 

I think one simple solution to this part is to not _fully_ deprecate the `volta` key within `package.json`, but to only use it for `node`/other versions that `packageManager` doesn't currently support, and probably for the "opt in to using `packageManager` key for version config" option.

Another alternative would be to use the `.node-version` file format that many other tools have used for quite a while, and continue deprecating the `volta` key in `package.json`

**Comment by dherman** [comment]
@0xdevalias I would actually advocate that approach. Well-defined precedence rules are typical for tools like Volta (see for example `rustup`'s docs). This creates a smooth way for us to evolve in the direction of using standard solutions when they're available, but still let Volta be out ahead when there is no established standard.

**Comment by donmccurdy** [comment]
This created some difficult-to-diagnose issues for me recently. I work in multiple repositories, some use Volta and some use Corepack. But when working in a project using Corepack (no `package.json#volta` set), Volta was silently disabling corepack, `corepack enable` did nothing with `package.json#packageManager`, and the wrong version of Yarn was chosen by Volta, leading to downstream confusion.

I'd thought I could have both Corepack and Volta enabled more-or-less safely, but that doesn't seem to be the case. It's not ideal to need to manually reconfigure my shell when switching subdirectories (one reason we use either tool!) but I think that is now necessary for me in any project using Volta, it'd be great to have a resolution on this. :)
